events/events.go

```diff
@@ -78,22 +79,46 @@ type Message interface {
 }
 
 type LaunchContext interface {
-	URL()         url.URL
-	Hashes()      map[string]string
-	Signatures()  map[string]string
+	URL() url.URL
+	Hashes() map[string]string
+	Signatures() map[string]string
 	ShortString() string
 }
 
+// this is copied from events to avoid cross reference
+type APISpecification struct {
+	SpecRef         string `json:"specRef"`         // A URL pointing to the definition of the API spec
+	Version         string `json:"version"`         // The version of the API spec in OSGI version format
+	ExclusiveAccess bool   `json:"exclusiveAccess"` // Whether or not exclusive access to this API spec is required
+	// API will allow. For a Consumer (agbot), this is likely to be 1.
+	// For a Producer, if this is zero, then no agreements, if 1 then
+	// then it's essentially exclusive access. For more than 1, then it's
+	// shared access. Added in version 2.
+	Arch string `json:"arch"` // The hardware architecture of the API spec impl. Added in version 2.
+}
+
+func (c APISpecification) String() string {
+	return fmt.Sprintf("SpecRef: %v, Version: %v, Arch: %v, ExclusiveAccess: %v", c.SpecRef, c.Version, c.Arch, c.ExclusiveAccess)
+}
+
+type MicroserviceSpec struct {
+	SpecRef     string
+	Version     string
+	Arch        string
+	InstanceKey string // key to the microservice instance table
+}
+
 type AgreementLaunchContext struct {
 	AgreementProtocol    string
 	AgreementId          string
 	Configure            ContainerConfig
 	ConfigureRaw         []byte
 	EnvironmentAdditions *map[string]string // provided by platform, not but user
+	Microservices        []MicroserviceSpec // for ms split.
```

Comment:
In governenace.go,  we got the microservices from the TsAndCs and put it in this struct and send it to container.go. container.go uses this check if the mcroservice containers are up and running or not before starting containers for workloads. My understanding is that this structure contains all that's needed to launch an agreement containers. It does not go to the db to get more info.